apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricSet.java

```diff
@@ -43,26 +44,73 @@
  * </pre>
  */
 public class MetricSet {
-    private final Map<String, String> labels;
-    private final ConcurrentMap<String, DoubleSupplier> samples = new ConcurrentHashMap<>();
+    private final Labels.Immutable labels;
+    private final ConcurrentMap<String, DoubleSupplier> gauges;
+    // low load factor as hash collisions are quite costly when tracking breakdown metrics
```

Comment:
Can this be improved by using a cache of immutable label objects for all combinations of the span types and subtypes we use, where equation is quickly done through reference comparison?

Reply:
Labels also contain the transaction name. If the transaction name is chosen poorly, it may have unlimited distinct values (for example if it contains an URL path parameter)

Reply:
Can the structure suggested in https://github.com/elastic/apm-agent-java/pull/588/files#r294691938 solve this?